swarm/src/lib.rs

```diff
@@ -651,19 +702,37 @@ where TBehaviour: NetworkBehaviour<ProtocolsHandler = THandler>,
 
             debug_assert!(this.pending_event.is_none());
 
+            // Assume the behaviour can make progress, unless determined otherwise.
+            this.behaviour_pending = false;
+
             let behaviour_poll = {
                 let mut parameters = SwarmPollParameters {
                     local_peer_id: &mut this.network.local_peer_id(),
                     supported_protocols: &this.supported_protocols,
                     listened_addrs: &this.listened_addrs,
                     external_addrs: &this.external_addrs
                 };
+                this.event_lead = this.event_lead.saturating_sub(1);
                 this.behaviour.poll(cx, &mut parameters)
             };
 
             match behaviour_poll {
-                Poll::Pending if network_not_ready => return Poll::Pending,
-                Poll::Pending => (),
+                Poll::Pending => {
+                    // The behaviour has nothing to produce, thus it is
+                    // waiting for more network activity and events.
+                    this.behaviour_pending = true;
+                    this.event_lead = 0;
+                    if network_pending {
+                        // The network was polled before polling the
+                        // behaviour and is waiting for I/O, hence there
+                        // is nothing further to do until the next wakeup.
+                        return Poll::Pending
+                    } else {
+                        // The network was not polled in this iteration and
```

Comment:
Given that not all `match this.network.poll(cx)` arms have a `return` this point could be reached even though the network was indeed polled, no? Or should one read this as `The network was not polled till it returned pending`?

E.g. the following match arm does not return:

```rust
Poll::Ready(NetworkEvent::ConnectionEvent { connection, event }) 
```

then if there is no pending event (`this.pending_event.take() == None`) and `this.behaviour.poll(cx, &mut parameters)` returns `Poll::Pending` one could end up here even though the network was polled.

Reply:
Right, I'll update the comment.